# Programming Assignment: Data preprocessing in Pandas 

This is the firts programimg assigment in the course. Еhe task is devoted to acquaintance with pandas.  
In the next paragraph, original six tasks will be presented in Russian.  

**Tasks will be duplicated in english further in this assignment**

Задачи:
1. Какое количество мужчин и женщин ехало на корабле? В качествеответа приведите два числа через пробел.

2. Какой части пассажиров удалось выжить? Посчитайте долю вы-живших пассажиров. Ответ приведите в процентах (число в ин-тервале от 0 до 100, знак процента не нужен), округлив до двухзнаков.

3. Какую долю пассажиры первого класса составляли среди всех пас-сажиров? Ответ приведите в процентах (число в интервале от 0 до100, знак процента не нужен), округлив до двух знаков.

4. Какого возраста были пассажиры? Посчитайте среднее и медиа-ну возраста пассажиров. Посчитайте среднее и медиану возрастапассажиров. В качестве ответа приведите два числа через пробел.

5. Коррелируют ли число братьев/сестер с числом родителей/детей?Посчитайте корреляцию Пирсона между признаками SibSp и Parch.

6. Какое самое популярное женское имя на корабле? Извлеките изполного имени пассажира (колонка Name) его личное имя (FirstName). 

Это задание типичный пример того, с чем сталкивается специалист по анализу данных. Данные очень разнородные ишумные, но из них требуется извлечь необходимую информацию.Попробуйте вручную разобрать несколько значений столбца Name и выработать правило для извлечения имен, а также разделенияих на женские и мужские

## Library import and data download

In [1]:
import pandas
import re

In [2]:
df = pandas.read_csv('titanic.csv',index_col='PassengerId')

We will be working with data from Titanic that was published on [kaggle](https://www.kaggle.com/c/titanic/overview)

In [3]:
df

Survived  Pclass  \
PassengerId                     
1                   0       3   
2                   1       1   
3                   1       3   
4                   1       1   
5                   0       3   
...               ...     ...   
887                 0       2   
888                 1       1   
889                 0       3   
890                 1       1   
891                 0       3   

                                                          Name     Sex   Age  \
PassengerId                                                                    
1                                      Braund, Mr. Owen Harris    male  22.0   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
5                                     Allen, Mr. William Henry    male  35.0   
...                                                        ...     ...   ...   
887                                      Montvila, Rev. Juozas    male  27.0   
888                               Graham, Miss. Margaret Edith  female  19.0   
889                   Johnston, Miss. Catherine Helen "Carrie"  female   NaN   
890                                      Behr, Mr. Karl Howell    male  26.0   
891                                        Dooley, Mr. Patrick    male  32.0   

             SibSp  Parch            Ticket     Fare Cabin Embarked  
PassengerId                                                          
1                1      0         A/5 21171   7.2500   NaN        S  
2                1      0          PC 17599  71.2833   C85        C  
3                0      0  STON/O2. 3101282   7.9250   NaN        S  
4                1      0            113803  53.1000  C123        S  
5                0      0            373450   8.0500   NaN        S  
...            ...    ...               ...      ...   ...      ...  
887              0      0            211536  13.0000   NaN        S  
888              0      0            112053  30.0000   B42        S  
889              1      2        W./C. 6607  23.4500   NaN        S  
890              0      0            111369  30.0000  C148        C  
891              0      0            370376   7.7500   NaN        Q  

[891 rows x 11 columns]

---

## Task №1
*Eng:* How many men and women were on the ship?

*Ru*: Какое количество мужчин и женщин ехало на корабле? В качествеответа приведите два числа через пробел.

In [4]:
df['Survived'].value_counts()

0    549
1    342
Name: Survived, dtype: int64

In [5]:
number_of_passengers = len(df)

In [6]:
len(df[df['Sex'] == 'male'])

577

In [7]:
len(df[df['Sex'] == 'female'])

314

577 mans and 314 womens according to data

## Task №2
*Eng:* How many passengers survived? Calculate the proportion of passengers who live. Give your answer as a percentage (a number in the range from 0 to 100, no percent sign is needed), rounded to two digits.

*Ru*: Какой части пассажиров удалось выжить? Посчитайте долю выживших пассажиров. Ответ приведите в процентах (число в интервале от 0 до 100, знак процента не нужен), округлив до двухзнаков.

In [8]:
number_of_survivors = len(df[df['Survived'] == 1])

In [9]:
number_of_survivors 

342

In [10]:
round(number_of_survivors * 100 / number_of_passengers, 2)

38.38

38.38% of passengers made it

## Task №3
*Eng:* What is the share of first class passengers among all passengers? Give your answer as a percentage (a number in the range from 0 to 100, no percent sign is needed), rounded to two digits. 

*Ru*: Какую долю пассажиры первого класса составляли среди всех пассажиров? Ответ приведите в процентах (число в интервале от 0 до100, знак процента не нужен), округлив до двух знаков.

In [11]:
round(len(df[df['Pclass'] == 1]) * 100/number_of_passengers, 2)

24.24

## Task №4
*Eng:* How old were the passengers? Calculate the average and median age of passengers. Give two numbers separated by spaces as your answer. 

*Ru*:
Какого возраста были пассажиры? Посчитайте среднее и медиану возраста пассажиров. В качестве ответа приведите два числа через пробел.

In [12]:
df['Age'].median(skipna = True)

28.0

In [13]:
df['Age'].mean(skipna = True)

29.69911764705882

## Task №5
*Eng:* Does the number of siblings correlate with the number of parents / children? Calculate the Pearson correlation between the SibSp and Parch traits. 

*Ru*:
Коррелируют ли число братьев/сестер с числом родителей/детей?Посчитайте корреляцию Пирсона между признаками SibSp и Parch.


In [14]:
df.corr().loc['SibSp','Parch']

0.41483769862015263

## Task №6
*Eng:* What's the most popular female name on a ship? Extract the passenger's personal name (FirstName) from the passenger's full name (Name column). 

*Ru*:
Какое самое популярное женское имя на корабле? Извлеките из полного имени пассажира (колонка Name) его личное имя (FirstName).

In [15]:
df['Name'][df['Sex'] == 'female'].max()

'de Messemaeker, Mrs. Guillaume Joseph (Emma)'

It would be too easy, the answer 'Emma' wasn't correct. So, I made little parser for this task.

In [16]:
data_name = df['Name'][df['Sex'] == 'female'].tolist()

In [17]:
lst = []
for name in data_name:
    if (name.find('Miss')!=-1):
        lst.append(name[name.find('Miss'):].split(' ')[1])
    else:
        if (name.find('Mrs')!=-1):
            if re.findall(r'\([ ?A-Za-z]+\)', name):
                lst.append(re.findall(r'\([ ?A-Za-z]+\)', name)[0].split(' ')[0].strip('()'))
            else:
                lst.append(name[name.find('Mrs'):].split(' ')[1].strip('()\"'))
        else: 
            if (name.find('Dr')!=-1):
                lst.append(name[name.find('Dr'):].split(' ')[1])
            else:
                if re.findall(r'\([ ?A-Za-z]+\)', name):
                    lst.append(re.findall(r'\([ ?A-Za-z]+\)', name)[0].split(' ')[0].strip('()'))
                else:
                    lst.append(name.split(' ')[-1].strip('(),'))
print(*lst)

Florence Laina Lily Elisabeth Adele Marguerite Elizabeth Hulda Mary Emelia Fatima Anna Torborg Selma Ellen Marie Mary Augusta Jamila Johanna Dorothy Simonne Margaret Bridget Josefine Myna Elizabeth Emily Constance Amelie Elizabeth Erna Lillian Elizabeth Brigdet Bertha Maria Mabel Ada Matilda Anna Bertha Hileni Katriina Malake Ellis Susan Anna Grace Mathilde Helen Sultana Anna Elin Robina Edith Katherine James Edith Anna Eleanor Ann Constance Luise Thomas Rosa Carla Margaret Elise Margaret Henriette Telma Helen Clear Madeleine Eliina Albina Mathilde Irene Lillian Alice Marjorie Thamine Katherine Agda Anna Elna Helena Hanne Gertrude Gladys Anna Lutie Delia Edith Amelia Elizabeth Hanora Kornelia Augusta Rosa Kate Ellen Helen Aloisia Helen Helene Anna Nora Margaret Maria Laura Margaret Emily Anna Helmina Miriam Mary Margaretta Hilda Sylvia Marie Ada Emily Jean Agnes Clara Elizabeth Alice Amelia Marion Mary Elsie Annie Mary Helen Catherine Anna Mantoura Annie Pauline Stina Leila Aurora Rosa

In [18]:
name_df = pandas.DataFrame(data = lst, columns=['Names'])
name_df

Names
0     Florence
1        Laina
2         Lily
3    Elisabeth
4        Adele
..         ...
309    Imanita
310      Gerda
311   Margaret
312   Margaret
313  Catherine

[314 rows x 1 columns]

In [19]:
name_df['Names'].value_counts().idxmax()

'Anna'